tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,341 @@
+while the requests for random bytes are served from the RNG. The RNG is used to
+"mix" the unpredictability inherent in its input, that is, to smooth out
+various timestamps and similar values into a statistically well-behaved
+sequence.
+
+## Do Virtual Machines Dream of Electric Sheep?
+
+The problem is that a virtual machine (VM) in a typical configuration barely
+sees any physical hardware. Users do not interact with VMs in server scenarios
+using a directly-connected keyboard and mouse. VMs make use of a virtualized
+network interface and virtualized disks. Even CPU features can be intercepted
+and virtualized. There is hardly any entropy in this environment.
+
+This is a known problem and [various][randomness-exposed]
+[analyses][not-so-random] of the weakness of RNGs in virtual environments have
+been published. The problem is especially severe in the beginning, that is,
+after boot. Since the gradual trickle of unpredictability from hardware events
+slowly moves the random stream into an unpredictable state, at the very start,
+it tends to be in a predictable state. Typically, operating systems store a
+block of their RNG output on shutdown and use it to quickly reseed the RNG on
+boot, in order to reuse the already-diverges state of the previous boot to
+quickly kick-start unpredictability. Unfortunately, it is common to boot several
+machines from the same system image, or from a pristine image lacking a RNG
+seed, making their RNGs vulnerable to prediction near to the startup phase.
+
+To help solve these problems, we employ several sources of entropy in MirageOS
+unikernels. In the case of a Unix executable, we simply reuse the system's RNG,
+as exposed via `/dev/urandom`, as the source of entropy. The kernel is in a much
+better position to enter an unpredictable state than a process running
+underneath. In the case of a unikernel running on top of Xen, we separate
+entropy sources according to their origin: some are within the unikernel itself,
+and some are external.
+
+In the external case, we again rely on the kernel interacting with the hardware
+(dom0) to harvest the actual environmental entropy. We have a background
+service, [Xentropyd][xentropyd], which runs in dom0, reads the RNG and serves
+its output to other domains through the Xen Console. The problem is that in many
+scenarios, like hosting on popular cloud providers, we cannot expect extra
+cooperation from dom0. (MAKE A NOTE: this mechanism is present, but not even
+activated in MirageOS 2.5.0!!)
+
+So we need to be able to achieve unpredictability relying purely on what is
+available inside a unikernel. As a unikernel does not directly see the hardware,
+to tap into the ambiental entropy, we have to continuously sample the
+inter-event timings. This process is analogous to what happens in a full-blown
+OS kernel, except our events are stripped of extra hardware context, and our
+timers are potentially less granular.
```

Comment:
Was not able to make much sense out from this paragraph. Which events are you using (I guess networking events) ? What do you mean by "our timers are potentially less granular" and what is the implication.